File: leetcode-62.py
from typing import List

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # dp = [1 for i in range(n)]
        # for i in range(1, m):
        #     for j in range(1, n):
        #         dp[j] = dp[j-1]+dp[j]
        # return dp[n-1]
        def fi(n, m):
            ans = 1
            while m > 0:
                ans *= n
                n -= 1
                m -= 1
            return ans
        a = m+n-2
        b = m-1
        ans = fi(a, b)//(fi(b,b))
        return ans

if __name__ == "__main__":
    so = Solution()
    print(so.uniquePaths(3,2))

File: leetcode/leetcode-62.cpp
/*
 *@File Name          :leetcode-62.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/7/21 9:44
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "algorithm"
#include "vector"
using namespace std;

class Solution {
 public:
  int uniquePaths(int m, int n) {
    int         a = m + n - 2;
    int         b = n - 1;
    vector<int> sb(b + 1, 0);
    int         k   = a - b;
    int         ans = 1;
    while (a > k) {
      ans *= a;
      a--;
      for (int i = b; i > 1; i--) {
        if (sb[i] == 0 && ans % i == 0) {
          ans /= i;
          sb[i] = 1;
        }
      }
    }
    return ans;
  }
};